apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+                    if (credentials.IsCreateScopedRequired)
+                    {
+                        credentials = credentials.CreateScoped(PublisherClient.DefaultScopes);
+                    }
+                    channelCredentials = credentials.ToChannelCredentials();
+                }
+                // Create the channels and clients, and register shutdown functions for each channel
+                var clients = new PublisherClient[clientCount];
+                var shutdowns = new Func<Task>[clientCount];
+                for (int i = 0; i < clientCount; i += 1)
+                {
+                    var channel = new Channel(PublisherClient.DefaultEndpoint.Host, PublisherClient.DefaultEndpoint.Port, channelCredentials);
+                    clients[i] = PublisherClient.Create(channel, settings?.PublisherSettings);
+                    shutdowns[i] = channel.ShutdownAsync;
+                }
+                shutdown = () => Task.WhenAll(shutdowns.Select(x => x()));
+                // Create new settings with the created clients, and clone the non-client settings
+                settings = new Settings(clients)
+                {
+                    BatchingSettings = settings.BatchingSettings,
+                    Scheduler = settings.Scheduler,
+                    BatchMaxElementCount = settings.BatchMaxElementCount,
+                    BatchMaxByteCount = settings.BatchMaxByteCount,
+                };
+            }
+            // Instantiate the publisher
+            return new HiPerfPublisherImpl(topicName, settings, shutdown, TaskHelper.Default);
+        }
+
+
+        /// <summary>
+        /// The associated <see cref="TopicName"/>. 
+        /// </summary>
+        public virtual TopicName TopicName => throw new NotImplementedException();
+
+        /// <summary>
+        /// Publish a message to the topic specified in <see cref="TopicName"/>.
+        /// </summary>
+        /// <param name="message">The <see cref="PubsubMessage"/> to publish.</param>
+        /// <returns>A task which completes once the message has been published.
+        /// The task <see cref="Task{String}.Result"/> contains the message ID.</returns>
+        public virtual Task<string> PublishAsync(PubsubMessage message) => throw new NotImplementedException();
+
+        /// <summary>
+        /// Publish a message to the topic specified in <see cref="TopicName"/>.
+        /// </summary>
+        /// <param name="message">The <see cref="PubsubMessage"/> to publish.</param>
+        /// <param name="encoding">The encoding for string to byte conversion.
+        /// If <c>null</c>, defaults to <see cref="Encoding.UTF8"/>.</param>
+        /// <returns>A task which completes once the message has been published.
+        /// The task <see cref="Task{String}.Result"/> contains the message ID.</returns>
+        public virtual Task<string> PublishAsync(string message, Encoding encoding = null) =>
```

Comment:
They seem to have accepted it as an idiom we'd like to maintain.
I guess it may come up for discussion again during beta...